c10/util/ArrayRef.h

```diff
@@ -22,10 +22,87 @@
 
 #include <array>
 #include <iterator>
+#include <memory>
+#include <type_traits>
 #include <vector>
 
 namespace c10 {
 
+
+
+struct TORCH_API SymbolicIntImpl {
+  bool virtual isSymbolicInt(int64_t data_) = 0;
+
+  // I can't overload + for primitive types
+  int64_t virtual add(int64_t s1, int64_t s2) = 0;
+  bool virtual equal(int64_t s1, int64_t s2) = 0;
+};
+
+
+class TORCH_API SymbolicOrConcreteInt {
+
+public:
+
+  SymbolicOrConcreteInt(int64_t d):
+  data_(d) {};
+
+  operator int64_t() {
+    return data_;
+  }
+
+  int64_t data_;
+
+  // Note this is static as it should apply to all SymbolicInts
```

Comment:
Why this limitation?
Wouldn't it be possible to re-interpret the `data_` into a `SymbolicInt*` and then call the method on that? c++ vtable will then do the magic to call into the implementation of whichever subclass this is.

Reply:
I would love to flush out this point actually. @ezyang  @zou3519 @wconstab 
I don't think @ezyang liked my global registry of SymbolicIntImpl either.
Initially, I was a bit concerned that adding `SymbolicInt` would overcomplicate memory management, but after talking to Ed, I don't think it might necessarily be the case tbh. 
I also had a concern that SymbolicOrConcreteInt to SymbolicInt conversions might be complicated, but it would probably look very similar to how I'm using `SymbolicIntImpl` anyways. 
I'm going to roughly prototype how it would look like shortly.